src/Psalm/Internal/Provider/ReturnTypeProvider/TriggerErrorReturnTypeProvider.php

```diff
@@ -0,0 +1,67 @@
+<?php declare(strict_types=1);
+
+namespace Psalm\Internal\Provider\ReturnTypeProvider;
+
+use Psalm\Plugin\EventHandler\Event\FunctionReturnTypeProviderEvent;
+use Psalm\Type;
+
+use function in_array;
+
+use const E_USER_DEPRECATED;
+use const E_USER_ERROR;
+use const E_USER_NOTICE;
+use const E_USER_WARNING;
+
+class TriggerErrorReturnTypeProvider implements \Psalm\Plugin\EventHandler\FunctionReturnTypeProviderInterface
+{
+    /**
+     * @return array<lowercase-string>
+     */
+    public static function getFunctionIds(): array
+    {
+        return ['trigger_error'];
+    }
+
+    public static function getFunctionReturnType(FunctionReturnTypeProviderEvent $event): ?Type\Union
+    {
+        $config = $event->getStatementsSource()->getCodebase()->config;
+        if ($config->trigger_error_behaviour === 'always') {
+            return new Type\Union([new Type\Atomic\TNever()]);
+        }
+
+        if ($config->trigger_error_behaviour === 'none') {
+            return new Type\Union([new Type\Atomic\TTrue()]);
+        }
+
+        //default behaviour
+        $call_args = $event->getCallArgs();
+        $statements_source = $event->getStatementsSource();
+        if (isset($call_args[1])
+            && ($array_arg_type = $statements_source->getNodeTypeProvider()->getType($call_args[1]->value))
+        ) {
+            $return_types = [];
+            foreach ($array_arg_type->getAtomicTypes() as $atomicType) {
+                if ($atomicType instanceof Type\Atomic\TLiteralInt) {
+                    if (in_array($atomicType->value, [E_USER_WARNING, E_USER_DEPRECATED, E_USER_NOTICE], true)) {
+                        $return_types[] = new Type\Atomic\TTrue();
+                    } elseif ($atomicType->value === E_USER_ERROR) {
+                        $return_types[] = new Type\Atomic\TNever();
+                    } else {
+                        // not recognized int literal. return false before PHP8, fatal error since
+                        $return_types[] = new Type\Atomic\TFalse();
+                    }
+                } elseif ($atomicType instanceof Type\Atomic\TInt) {
+                    // not specific integer. Returning a mix of never and bool would not make sense for Psalm
+                    $return_types[] = new Type\Atomic\TMixed();
+                } else {
+                    $return_types[] = new Type\Atomic\TMixed();
+                }
+            }
+
+            return new Type\Union($return_types);
```

Comment:
Shouldn't we first simplify `never|bool` to just `bool`, etc? `Union` doesn't do that automatically. There should be an existing method for that, but specific name escapes me at the moment.

Reply:
> simplify never|bool to just bool

is it really equivalent? when never is a part of the type, it means that sometimes there will be a crash at runtime if we try to return the result

Reply:
*If* `trigger_error()` returns, its return value is `bool`. If it doesn't (`never`) then there's no return value.

Reply:
you're right. I confused errors from psalm with runtime. I'll fix that

Reply:
>  but specific name escapes me at the moment.

I think it's `TypeCombiner::combine()`